File: Python/ds/Palindrome-doublylinkedlist.py
'''

Palindrome : A sequence which appears the same when read forwards and backwards
Purpose : To find if the given string is a palindrome or not, using doubly linked list

Method : Data Structures :- Doubly Linked List
        - Create a doubly linked list
        - Assign each character to one node of the doubly linked list
        - Check if the data in the before node and after node are equal
        - If they are equal, perform increment and decrement operations
        - If they are not equal, return false

Time complexity : 0(n)
Space complexity : 0(1)
Arguments : Arbitrary arguments
Return : Message if the input is a palindrome or not
Sample Input/Output :

- Test case 1:
Enter the string : clouds
Not a Palindrome

- Test case 2:
Enter the string : madam
Is a Palindrome

- Test case 3:
Enter the string : books
Not a Palindrome

'''


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoublyLinkedList:

    @staticmethod
    def printList(node):

        '''Reversing the string'''
        rev_key = ""
        while node:
            rev_key += node.data
            node = node.next
        return rev_key

    def __init__(self):
        self.head = None

    def push(self, new_data):

        ''' Adding new node in the front of the list '''
        new_node = Node(new_data)

        if self.head is None:
            self.head = new_node
            return

        last = self.head

        while last.next:
            last = last.next

        last.next = new_node
        new_node.prev = last
        return


if __name__ == "__main__":

    llist = DoublyLinkedList()
    keys = input("Enter the string : ")

    for i in reversed(range(len(keys))):
        llist.push(keys[i])

    rev_keys = llist.printList(llist.head)

    if keys == rev_keys:
        print("Is Palindrome")
    else:
        print("Not a Palindrome")

File: C/ds/palindrome_doubly_linked_list.c
/*
DOUBLY LINKED LIST TO CHECK IF A STRING IS PALINDROME
A string is accepted as input from the user.
The characters in the string are inserted as nodes
into the doubly linked list.
Two pointers lptr and rptr are used to point to the
beginning and end of the string respectively.
The list is traversed from left to right using lptr
and right to left using rptr.
If the characters match until they reach the same node
the string is palindrome
*/

#include <stdio.h>
#include <stdlib.h>

//structure for doubly linked list
typedef struct node
{
    char data;
    struct node *rlink;
    struct node *llink;
} node;

//to insert into doubly linked list
void insert(node *head, char a)
{
    node *n = malloc(sizeof(node));
    n->data = a;
    n->rlink = NULL;
    node *ptr = head;
    while (ptr->rlink)
        ptr = ptr->rlink;
    ptr->rlink = n;
    n->llink = ptr;
}

//to check if the string is palindrome
int palindrome(node *head)
{
    if (head->rlink == NULL)
        return 1;
    node *rptr = head;
    while (rptr->rlink)
        rptr = rptr->rlink;
    node *lptr = head->rlink;
    while (lptr != rptr)
    {
        if (lptr->data != rptr->data)
            return 0;
        lptr = lptr->rlink;
        rptr = rptr->llink;
    }
    return 1;
}

//driver code
void main()
{
    node *head = malloc(sizeof(node));
    char str[30];
    printf("Enter a string : ");
    scanf("%s", str);
    for (int i = 0; str[i] != '\0'; i++)
        insert(head, str[i]);
    if (palindrome(head) == 1)
        printf("%s is a palindrome\n", str);
    else
        printf("%s is not a palindrome\n", str);
}

/*
SAMPLE I/O:
Enter a string : refer
refer is a palindrome

Time Complexity: O(n)
Space Complexity: O(n)
*/
